apis/projectcontour/v1alpha1/contourconfig.go

```diff
@@ -22,6 +25,396 @@ import (
 // It contains most of all the options that can be customized, the
 // other remaining options being command line flags.
 type ContourConfigurationSpec struct {
+	// XDSServer contains parameters for the xDS server.
+	// +optional
+	XDSServer ServerParameters `json:"xdsServer,omitempty"`
+
+	// IngressOptions contains parameters for ingress options.
+	// +optional
+	IngressOptions Ingress `json:"ingress,omitempty"`
+
+	EnvoyConfig Envoy `json:"envoy,omitempty"`
+
+	// GatewayConfig contains parameters for the gateway-api Gateway that Contour
+	// is configured to serve traffic.
+	// +optional
+	GatewayConfig *GatewayParameters `json:"gateway,omitempty"`
+
+	// HTTPProxy defines parameters on HTTPProxy.
+	// +optional
+	HTTPProxy *HTTPProxyConfig `json:"httpproxy,omitempty"`
+
+	// LeaderElection contains leader election parameters.
+	// +optional
+	LeaderElection LeaderElectionParameters `json:"leaderelection,omitempty"`
+
+	// EnableExternalNameService allows processing of ExternalNameServices
+	// Defaults to disabled for security reasons.
+	// +optional
+	EnableExternalNameService bool `json:"enableExternalNameService,omitempty"`
+
+	// RateLimitService optionally holds properties of the Rate Limit Service
+	// to be used for global rate limiting.
+	RateLimitService RateLimitService `json:"rateLimitService,omitempty"`
+
+	// Policy specifies default policy applied if not overridden by the user
+	Policy PolicyParameters `json:"policy,omitempty"`
+}
```

Comment:
There's a mix of "[area]Config", "[area]Parameters", "[area]Options", and just "[area]" in the struct and field names, it'd be nice to be consistent. Perhaps we can just go with e.g. `LeaderElection`, `Gateway`, etc. for field names (since they're already within a `ContourConfigurationSpec` struct so repeating `Config` is probably redundant)? Struct names could either match the field names, or be e.g. `LeaderElectionConfig`, `GatewayConfig`, etc.